transport.go

```diff
@@ -24,19 +40,27 @@ type Conn interface {
 // but many more can be implemented, sctp, audio signals, sneakernet, UDT, a
 // network of drones carrying usb flash drives, and so on.
 type Transport interface {
-	Dialer(laddr ma.Multiaddr, opts ...DialOpt) (Dialer, error)
+	// Dial dials a remote peer. It should try to reuse local listener
+	// addresses if possible but it may choose not to.
+	Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (Conn, error)
```

Comment:
Given that we're breaking things, I got rid of the distinction between `Dial` and `DialContext`. Just pass a context.

Also, this now takes a peer ID. While we can technically get away with not having this, it allows us to validate connections early (given that transport security now happens *inside* the transport).